### PR TITLE
Taxi edge turn radius (i.e. smooth curve on each turn)

### PR DESCRIPTION
**Pull request by Adam-Shea**
Associated issues: This is a follow up to this discussion [#3135](https://github.com/cytoscape/cytoscape.js/discussions/3135)

Added rounded corners to haystacks when the `taxi-curve` property is set to yes.
image
Happy to change property name or move elsewhere if you think there is a better format
Thanks :)

Author:

- [x] The proper base branch has been selected.  New features go on `unstable`.  Bug-fix patches can go on either `unstable` or `master`.
- [x] Automated tests have been included in this pull request, if possible, for the new feature(s) or bug fix.  Check this box if tests are not pragmatically possible (e.g. rendering features could include screenshots or videos instead of automated tests).
- [x] The associated GitHub issues are included (above).
- [x] Notes have been included (above).

Reviewers:

- [ ] All automated checks are passing (green check next to latest commit).
- [ ] At least one reviewer has signed off on the pull request.
- [ ] For bug fixes:  Just after this pull request is merged, it should be applied to both the `master` branch and the `unstable` branch.  Normally, this just requires cherry-picking the corresponding merge commit from `master` to `unstable` -- or vice versa.

### CODE REVIEW

File: src/extensions/renderer/canvas/drawing-edges.js [hunk 1]
```diff
@@ -207,9 +207,18 @@ CRp.drawEdgePath = function( edge, context, pts, type ){
 
       case 'straight':
       case 'segments':
-      case 'haystack':
```
Comment (maxkfranz) [inline src/extensions/renderer/canvas/drawing-edges.js:210]: Why has this change been made re. haystacks?

File: src/style/properties.js [hunk 2]
```diff
@@ -349,6 +349,7 @@ const styfn = {};
     { name: 'taxi-turn', type: t.bidirectionalSizeMaybePercent, triggersBounds: diff.any },
     { name: 'taxi-turn-min-distance', type: t.size, triggersBounds: diff.any },
     { name: 'taxi-direction', type: t.axisDirection, triggersBounds: diff.any },
+    { name: 'taxi-curve', type: t.bool, triggersBounds: diff.any },
```
Comment (maxkfranz) [inline src/style/properties.js:352]: The first thing other people who would use this feature would ask is:  How do I set the radius of the curve, [analogous to `border-radius` in CSS](https://developer.mozilla.org/en-US/docs/Web/CSS/border-radius)?

With a small radius, the effect of the PR is nearly indistinguishable from the existing, [curved line-join](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/lineJoin).  Some people will certainly want a subtle curve, but many people will want a stronger one.

With a configurable radius, you could get away with using a single property (e.g. `taxi-turn-radius`, default 0).  In that case, 0 would be equivalent to your proposed `taxi-curve: no`.

**Review by maxkfranz — COMMENTED** [review]
This looks like a good start, and I'm sure it would be mergeable with some iteration.

**Comment by maxkfranz** [comment]
It would be useful to specify exactly how you propose your feature should be used.  Give an example of how the style properties, old and new, should be specified in order to achieve your proposed effect.  For instance, it's not clear why this would be coupled to haystack edges.

**Comment by hassniazi** [comment]
Great feedback, thanks @maxkfranz - will take a look and propose some changes

**Comment by maxkfranz** [comment]
Great.  We're planning on releasing 3.26.0 on next week's Friday, the 4th of August.  It would be great to have this feature in the release.

Looking forward to your updates.

**Comment by hassniazi** [comment]
@maxkfranz - we've made most of the changes now to get this feature in. Just in the process of testing / fixing up, will get the PR updated shortly...

**Comment by Adam-Shea** [comment]
Hey, I've updated the pr with the suggestions you made.
Now when supplying `taxi-radius`, you can set a custom curve for your taxi corners.
I've fix the haystack removal, that was an issue with my merge sorry. Now this sits within haystack as taxi uses it by default. I have tested both with and without taxi-radius and taxi works just fine.
Happy to make anymore suggestions.

File: src/extensions/renderer/canvas/drawing-edges.js [hunk 3]
```diff
@@ -208,8 +208,18 @@ CRp.drawEdgePath = function( edge, context, pts, type ){
       case 'straight':
       case 'segments':
       case 'haystack':
-        for( let i = 2; i + 1 < pts.length; i += 2 ){
-          context.lineTo( pts[ i ], pts[ i + 1] );
+        const taxiRadius= edge.pstyle( 'taxi-radius' )
+        const taxiRadiusValue = taxiRadius ? taxiRadius.value : 0
```
Comment (maxkfranz) [inline src/extensions/renderer/canvas/drawing-edges.js:212]:
```suggestion
        const taxiRadiusValue = taxiRadius ? taxiRadius.pfValue : 0
```

Comment (maxkfranz) [reply src/extensions/renderer/canvas/drawing-edges.js:212]: Preferred value (`pfValue`, i.e. px), not raw value (`value`, e.g. `1em` has `value` of `1`)

File: src/extensions/renderer/canvas/drawing-edges.js [hunk 4]
```diff
@@ -208,8 +208,18 @@ CRp.drawEdgePath = function( edge, context, pts, type ){
...
+
+        if( taxiRadiusValue > 0) {
```
Comment (maxkfranz) [inline src/extensions/renderer/canvas/drawing-edges.js:214]: This block should only apply to taxi edges, or the property should explicitly support all the affected cases.  

Currently, it's breaking other edges.  E.g. on the debug page with all edges set with `taxi-radius: 32px`:

Expected:

Screenshot 2023-08-01 at 08 46 33

Actual:

Screenshot 2023-08-01 at 08 46 18

File: src/style/properties.js [hunk 5]
```diff
@@ -349,6 +349,7 @@ const styfn = {};
     { name: 'taxi-turn', type: t.bidirectionalSizeMaybePercent, triggersBounds: diff.any },
     { name: 'taxi-turn-min-distance', type: t.size, triggersBounds: diff.any },
     { name: 'taxi-direction', type: t.axisDirection, triggersBounds: diff.any },
+    { name: 'taxi-radius', type: t.number, triggersBounds: diff.any },
```
Comment (maxkfranz) [inline src/style/properties.js:352]: Great name.  It should be documented: https://github.com/cytoscape/cytoscape.js/blob/unstable/documentation/md/style.md

File: src/extensions/renderer/canvas/drawing-edges.js [hunk 6]
```diff
@@ -208,8 +208,18 @@ CRp.drawEdgePath = function( edge, context, pts, type ){
...
+          for( let i = 2; i + 1 < pts.length; i += 2 ) {
+            context.arcTo( pts[ i - 2 ], pts[ i - 1 ], pts[ i ], pts[ i + 1 ], 2 * taxiRadiusValue )
+          }
+          context.lineTo( pts[ pts.length - 2 ], pts[ pts.length - 1 ] );
```
Comment (maxkfranz) [inline src/extensions/renderer/canvas/drawing-edges.js:218]: It looks like the hit tests aren't working on the curves.  On the debug page (`npm run watch`):

https://github.com/cytoscape/cytoscape.js/assets/989043/9aed9253-c6da-4728-b5db-17471f60eb46

Comment (maxkfranz) [reply src/extensions/renderer/canvas/drawing-edges.js:218]: The hit tests are here:

https://github.com/cytoscape/cytoscape.js/blob/a653d1b073bf65f24ae69db5927fc88492f34293/src/extensions/renderer/base/coord-ele-math/coords.js#L79-L318

File: src/extensions/renderer/canvas/drawing-edges.js [hunk 7]
```diff
@@ -208,8 +208,18 @@ CRp.drawEdgePath = function( edge, context, pts, type ){
...
+        } else {
+          for( let i = 2; i + 1 < pts.length; i += 2 ) {
+            context.lineTo( pts[ i ], pts[ i + 1] );
+          }
```
Comment (maxkfranz) [inline src/extensions/renderer/canvas/drawing-edges.js:222]: Note:  Taxi edges have an 'edge type' of 'segments', since they are currently rendered in the same way as segments -- but with the intermediate points calculated for you rather than specified manually.

**Review by maxkfranz — CHANGES_REQUESTED** [review]
Great.  Getting close.

**Comment by hassniazi** [comment]
We've pushed up the latest changes @maxkfranz - let us know if you need anything else

File: src/extensions/renderer/base/coord-ele-math/coords.js [hunk 8]
```diff
@@ -169,13 +169,37 @@ BRp.findNearestElements = function( x, y, interactiveElementsOnly, isTouch ){
...
 
       for( var i = 0; i + 3 < pts.length; i += 2 ){
-        if(
-          (math.inLineVicinity( x, y, pts[ i ], pts[ i + 1], pts[ i + 2], pts[ i + 3], width2 ))
-            &&
-          widthSq > ( sqDist = math.sqdistToFiniteLine( x, y, pts[ i ], pts[ i + 1], pts[ i + 2], pts[ i + 3] ) )
-        ){
-          addEle( edge, sqDist );
-          return true;
+        if( rs.edgeType === 'segments' && edge.pstyle( 'taxi-radius' ).value > 0 ) {
```
[the remark this answers is not in the document]
Comment (maxkfranz) [reply src/extensions/renderer/base/coord-ele-math/coords.js:172]: Also:  Hit test code that is curved-taxi-specific should check `edge.pstyle( 'curve-style' ).value === 'taxi'`, just like your drawing code, so that it's explicit and bounded.

File: documentation/md/style.md [hunk 9]
```diff
@@ -456,7 +456,10 @@ When a taxi edge would be impossible to draw along the regular turning plan ---
   * Note that bundling may not work with an explicit direction (`upward`, `downward`, `leftward`, or `rightward`) in tandem with a turn distance specified in percent units.
 * **`taxi-turn-min-distance`** : The minimum distance along the primary axis that is maintained between the nodes and the turns.
   * This value only takes on absolute values (e.g. `5px`).
-  * This property makes the taxi edge be re-routed when the turns would be otherwise too close to the source or target.  As such, it also helps to avoid turns overlapping edge endpoint arrows.
+  * This property makes the taxi edge be re-routed when the turns would be otherwise too close to the source or target.  As such, it also helps to avoid turns overlapping edge endpoint arrows
+* **`taxi-radius`** : The taxi-radius property applies a curve to a taxi edge, providing a rounded appearance to any 90 degree bend.
+  * This value should be a floating point number (e.g. `3.14`).
+
```
Comment (maxkfranz) [inline documentation/md/style.md:462]: Even if we cap the radius (see the comments later on), it may be useful to have another sub-bullet-point that notes that the radius should not be too large.  It depends on how your other taxi properties are specified, the layout, etc.

File: src/extensions/renderer/canvas/drawing-edges.js [hunk 10]
```diff
@@ -208,8 +208,18 @@ CRp.drawEdgePath = function( edge, context, pts, type ){
       case 'straight':
       case 'segments':
       case 'haystack':
-        for( let i = 2; i + 1 < pts.length; i += 2 ){
-          context.lineTo( pts[ i ], pts[ i + 1] );
+        const taxiRadius= edge.pstyle( 'taxi-radius' )
+        const taxiRadiusValue = taxiRadius ? taxiRadius.pfValue : 0
```
Comment (maxkfranz) [inline src/extensions/renderer/canvas/drawing-edges.js:212]: Similar to how `border-radius` works in ordinary CSS, it would be useful to cap the taxi radius based on the available space.  For instance, a div that's a 10px by 10px square can't have a radius greater than 5px effectively.

For the edges, the radius could be capped by the available space on the line.

If you have a larger radius, then the edge will bug out depending on the orientation of the nodes:


https://github.com/cytoscape/cytoscape.js/assets/989043/4fa3d97b-83fc-4b51-99a9-a082c33d2a01

**Review by maxkfranz — CHANGES_REQUESTED** [review]
Great.  We're making progress.  I've attached some comments below:

File: documentation/md/style.md [hunk 11]
```diff
@@ -456,7 +456,10 @@ When a taxi edge would be impossible to draw along the regular turning plan ---
   * Note that bundling may not work with an explicit direction (`upward`, `downward`, `leftward`, or `rightward`) in tandem with a turn distance specified in percent units.
 * **`taxi-turn-min-distance`** : The minimum distance along the primary axis that is maintained between the nodes and the turns.
   * This value only takes on absolute values (e.g. `5px`).
-  * This property makes the taxi edge be re-routed when the turns would be otherwise too close to the source or target.  As such, it also helps to avoid turns overlapping edge endpoint arrows.
+  * This property makes the taxi edge be re-routed when the turns would be otherwise too close to the source or target.  As such, it also helps to avoid turns overlapping edge endpoint arrows
+* **`taxi-radius`** : The taxi-radius property applies a curve to a taxi edge, providing a rounded appearance to any 90 degree bend.
+  * This value should be a floating point number (e.g. `3.14`).
```
[the remark this answers is not in the document]
Comment (maxkfranz) [reply documentation/md/style.md:461]: Just a clarification

File: src/extensions/renderer/canvas/drawing-edges.js (hunk 10 above)
Comment (maxkfranz) [reply src/extensions/renderer/canvas/drawing-edges.js:212]: N.b. this issue also effects smaller radii, but not as obviously.

**Comment by Adam-Shea** [comment]
Heya, all clear! Just finding some time to get the last few changes in

**Comment by maxkfranz** [comment]
@Adam-Shea, are you still interested in this feature?
